### ISSUE TITLE
Invalid `do:switch`

### ISSUE THREAD

**Issue by JBBianchi** [serverlessworkflow/specification#894]
**What seems off**:
After #875, and because of, https://github.com/serverlessworkflow/specification/discussions/884#discussioncomment-9662801, #882 was merged, making `do` a single task entry. This breaks the possibility to use `do:switch` as there is nothing to switch to. It forces the user to use `do:execute:sequentially:switch`. 

**What you expected to be**:
We might need to rollback to the original idea of #875, which is to have `do` as `map[string, task]`, but not only for top level.

**Anything else we need to know?**:
As mentioned by @matthias-pichler-warrify in #875, making `do` a `map[string, task]`, the `execute.sequentially` directive doesn't really make sense anymore. We might want to refactor `execute` at the same time (or keep that for another issue).

**Comment by fjtirado** [comment]
Can we do this for sequentiality

```
do:
   -task1:
    -task2: 
```

and this for concurrency

```
do:
   -task1:
   -task2: 
onParallel:
   compete: false
```

If we agree, thats ok, I think that will be doable schema wise (by putting do: as required and onParallel: as optional of a new schema construct)

**Comment by JBBianchi** [comment]
Discussion from #875 related to this issue:

@JBBianchi https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158650129
> I think we overlooked the case of `switch` with the adopted solution. It's relegated to a second zone citizen, only available in `execute`:
> 
> ```yaml
> do:
>   execute:
>     sequentially:
>       - processOrder:
>           switch:
>           # ...
> ```
> 
> https://github.com/serverlessworkflow/specification/blob/main/dsl-reference.md#switch
> 
> Shouldn't we go back to `do` being a `map[string, task][]` instead of a single task ?

@cdavernas https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158657771
> @JBBianchi Yes, indeed. Do as a mapping array is what we should opt for:
> 
>     * It resolves consistency of the keyword `do`: it's alway a mapping array of tasks
> 
>     * It allows for top-level switch statements
> 
>     * It's more consistent with what a workflow is, to most people: a successions of multiple tasks, not a single top level one, which is then... a task

@matthias-pichler-warrify https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158697288
> That is true. And `do` as an array is absolutely fine by me, however as @cdavernas pointed out: what's the purpose of `sequentially` then?
> 
> Which given the discussion around `concurrently` and the `execute` keyword in general, might mean:
> 
>     * drop sequentially (not needed as we have arrays everywhere
> 
>     * only keep `concurrently` (or replace with `branch` to make it a verb for consistency)

@cdavernas https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158732278
> > multitask
> > /ˈmʌltɪtɑːsk/
> > verb
> > 
> > 1. (of a person) deal with more than one task at the same time.
> >    "I managed my time efficiently and multitasked"
> 
> It is a verb, though certainly not the best term. But there is only a little amount of alternatives, such as the (horrible) `parallelize`
> 
> > The idea for branch also came from AWS StepFunctions (they run in parallel there)
> 
> Interesting. Might be worth digging.

@JBBianchi https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158762394
> Do you mean:
> 
> ```yaml
> do:
>   execute:
>     branch:
>       - task1:
>       - task2:
>       - task3:    
> ```
> 
> or
> 
> ```yaml
> do:
>   branch:
>     - task1:
>     - task2:
>     - task3: 
> ```
> 
> (an alternative to `branch` could also be `fork`)
> 
> In the first case, we keep the "controversial" `execute` but have an extra level to put config like `compete: true`. In the second one, we get rid of the controversial keyword but we might need an extra level of depth.

@fjtirado https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158878743
> @JBBianchi Im trying to undestand the issue with swith. I understood that rather than writing
> 
> ```
> do:
>   execute:
>     sequentially:
>       - processOrder:
>           switch:
>             - case1:
>                 when: .orderType == "electronic"
>                 then: processElectronicOrder
>             - case2:
>                 when: .orderType == "physical"
>                 then: processPhysicalOrder
>             - default:
>                 then: handleUnknownOrderType
>       - processElectronicOrder:
>           execute:
>             sequentially:
>               - validatePayment: {}
>               - fulfillOrder: {}
>           then: exit
>       - processPhysicalOrder:
>           execute:
>             sequentially:
>               - checkInventory: {}
>               - packItems: {}
>               - scheduleShipping: {}
>           then: exit
>       - handleUnknownOrderType:
>           execute:
>             sequentially:
>               - logWarning: {}
>               - notifyAdmin: {}
> ```
> 
> we should be able to write
> 
> ```
> do:
>      - processOrder:
>           switch:
>             - case1:
>                 when: .orderType == "electronic"
>                 then: processElectronicOrder
>             - case2:
>                 when: .orderType == "physical"
>                 then: processPhysicalOrder
>             - default:
>                 then: handleUnknownOrderType
>       - processElectronicOrder:
>           execute:
>             sequentially:
>               - validatePayment: {}
>               - fulfillOrder: {}
>           then: exit
>       - processPhysicalOrder:
>           execute:
>             sequentially:
>               - checkInventory: {}
>               - packItems: {}
>               - scheduleShipping: {}
>           then: exit
>       - handleUnknownOrderType:
>           execute:
>             sequentially:
>               - logWarning: {}
>               - notifyAdmin: {}
> ```
> 
> isnt it? But it we remove execute, as option 2 in this [discussion](https://github.com/serverlessworkflow/specification/discussions/889), it would be
> 
> ```
> do:
>   sequentially:
>      - processOrder:
>           switch:
>             - case1:
>                 when: .orderType == "electronic"
>                 then: processElectronicOrder
>             - case2:
>                 when: .orderType == "physical"
>                 then: processPhysicalOrder
>             - default:
>                 then: handleUnknownOrderType
>       - processElectronicOrder:
>             sequentially:
>               - validatePayment: {}
>               - fulfillOrder: {}
>           then: exit
>       - processPhysicalOrder:
>             sequentially:
>               - checkInventory: {}
>               - packItems: {}
>               - scheduleShipping: {}
>           then: exit
>       - handleUnknownOrderType:
>            sequentially:
>               - logWarning: {}
>               - notifyAdmin: {}
> ```
> 
> Am I missing anything? (probably I am, my apologies in advance)

@JBBianchi https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2158952423
> > @JBBianchi Im trying to undestand the issue with swith. I understood that rather than writing
> > ```
> > do:
> >   execute:
> >     sequentially:
> >       - processOrder:
> >           switch:
> >             - case1:
> >                 when: .orderType == "electronic"
> >                 then: processElectronicOrder
> >             - case2:
> >                 when: .orderType == "physical"
> >                 then: processPhysicalOrder
> >             - default:
> >                 then: handleUnknownOrderType
> >       - processElectronicOrder:
> >           execute:
> >             sequentially:
> >               - validatePayment: {}
> >               - fulfillOrder: {}
> >           then: exit
> >       - processPhysicalOrder:
> >           execute:
> >             sequentially:
> >               - checkInventory: {}
> >               - packItems: {}
> >               - scheduleShipping: {}
> >           then: exit
> >       - handleUnknownOrderType:
> >           execute:
> >             sequentially:
> >               - logWarning: {}
> >               - notifyAdmin: {}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > we should be able to write
> > ```
> > do:
> >      - processOrder:
> >           switch:
> >             - case1:
> >                 when: .orderType == "electronic"
> >                 then: processElectronicOrder
> >             - case2:
> >                 when: .orderType == "physical"
> >                 then: processPhysicalOrder
> >             - default:
> >                 then: handleUnknownOrderType
> >       - processElectronicOrder:
> >           execute:
> >             sequentially:
> >               - validatePayment: {}
> >               - fulfillOrder: {}
> >           then: exit
> >       - processPhysicalOrder:
> >           execute:
> >             sequentially:
> >               - checkInventory: {}
> >               - packItems: {}
> >               - scheduleShipping: {}
> >           then: exit
> >       - handleUnknownOrderType:
> >           execute:
> >             sequentially:
> >               - logWarning: {}
> >               - notifyAdmin: {}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > isnt it?
> 
> That's my concern indeed.
> 
> > But it we remove execute, as option 2 in this [discussion](https://github.com/serverlessworkflow/specification/discussions/889), it would be
> > ```
> > do:
> >   sequentially:
> >      - processOrder:
> >           switch:
> >             - case1:
> >                 when: .orderType == "electronic"
> >                 then: processElectronicOrder
> >             - case2:
> >                 when: .orderType == "physical"
> >                 then: processPhysicalOrder
> >             - default:
> >                 then: handleUnknownOrderType
> >       - processElectronicOrder:
> >           execute:
> >             sequentially:
> >               - validatePayment: {}
> >               - fulfillOrder: {}
> >           then: exit
> >       - processPhysicalOrder:
> >           execute:
> >             sequentially:
> >               - checkInventory: {}
> >               - packItems: {}
> >               - scheduleShipping: {}
> >           then: exit
> >       - handleUnknownOrderType:
> >           execute:
> >             sequentially:
> >               - logWarning: {}
> >               - notifyAdmin: {}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Am I missing anything? (probably I am, my apologies in advance)
> 
> The later example doesn't seem to match any late suggestions made in this issue. I didn't mention removing `execute` per se as it's another topic. But if I understand what @matthias-pichler-warrify and @cdavernas were talking about just after I raised my concern, is that `sequentially` is the default behavior of a `do`, therefore `execute.sequentially` doesn't really have a reason to exists. Which leaves us with `concurrently`, which could be replaced by `branch`(, `multitask`, `parallelize` or `fork`). What's unclear is if we keep `execute.branch` (or alternative) or we just get rid of `execute` altogether an opt for `branch` (or alternative) straight on - which then, rejoin the point 2 of #889 and makes the discussion obsolete. In the later case though, we'd probably need an extra level of depth for some params like racing (or threads count for instance).
> 
> e.g.:
> 
> ```yaml
> do:
>   - processOrder:
>       switch:
>         - case1:
>             when: .orderType == "electronic"
>             then: processElectronicOrder
>         - case2:
>             when: .orderType == "physical"
>             then: processPhysicalOrder
>         - default:
>             then: handleUnknownOrderType
>   - processElectronicOrder:
>       do:
>         - validatePayment: {}
>         - fulfillOrder: {}
>         then: exit
>   - processPhysicalOrder:
>       execute:
>         compete: true
>         branch:  # `multitask`, `parallelize` or `fork`
>           - checkInventory: {}
>           - packItems: {}
>           - scheduleShipping: {}
>             then: exit
>   - handleUnknownOrderType: #...
> ```
> 
> or
> 
> ```yaml
> do:
>   - processOrder:
>       switch:
>         - case1:
>             when: .orderType == "electronic"
>             then: processElectronicOrder
>         - case2:
>             when: .orderType == "physical"
>             then: processPhysicalOrder
>         - default:
>             then: handleUnknownOrderType
>   - processElectronicOrder:
>       do:
>         - validatePayment: {}
>         - fulfillOrder: {}
>         then: exit
>   - processPhysicalOrder:
>       branch:
>         compete: true
>         on:  # `multitask`, `parallelize` or `fork`
>           - checkInventory: {}
>           - packItems: {}
>           - scheduleShipping: {}
>             then: exit
>   - handleUnknownOrderType: #...
> ```
> 
> @cdavernas @matthias-pichler-warrify am I getting it right ?

@fjtirado https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2160115152
> @JBBianchi There was a typo in my last example, I edited it. My point was that this is kind of related with the solution for do: execute:, if we simplify how multitask is written, its not really an issue that switch is a task which only makes sense within a multitask (which is why we are proposing to make do: multitask again) Also, doing do: always a multitask (so if you want a single task you just write one named task, as far as this also applies to loops, retries and catch) and removing sequentially (because its implicit in do:), leaving concurrently as an special case of multitask will also work.

@cdavernas https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2160130745
> > its not really an issue that switch is a task which only makes sense within a multitask
> 
> I strongly disagree: it is a problem, and a big one. The fact that `do` could have different meaning is blocking for you in terms of consistency, but not the fact that the `switch` task - and the `switch` task only - cannot be used at top level? Furthermore, as said above, a single task per workflow goes against the definition of a workflow. It also is a problem when it comes to visualization.
> 
> > (which is why we are proposing to make do: multitask again)
> 
> That is not what we are proposing. We are proposing:
> 
>     1. To make `do` a (sequential) array everywhere
> 
>     2. To get rid of `execute/sequentially`
> 
>     3. Create a new task to "parallelize" tasks

@fjtirado https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2160144367
> @cdavernas I think we are are saying the same thing, so there is not need to strongly disagree ;) When I said that we are proposing to do a multitask I mean do: being an array (which I said is fine as far as do: have the same syntax everywhere, not only in main flow) And removing execute/sequentially is part of the do:execute: redundancy discussion.

@cdavernas https://github.com/serverlessworkflow/specification/issues/875#issuecomment-2160148819
> > I think we are are saying the same thing, so there is not need to strongly disagree ;)
> 
> Hehe, sorry, seems I misunderstood you indeed 🤣
> 
> > And removing execute/sequentially is part of the do:execute: redundancy discussion.
> 
> Indeed.

**Comment by matthias-pichler-warrify** [comment]
@JBBianchi with this example:

```yaml
do:
  - processOrder:
      switch:
        - case1:
            when: .orderType == "electronic"
            then: processElectronicOrder
        - case2:
            when: .orderType == "physical"
            then: processPhysicalOrder
        - default:
            then: handleUnknownOrderType
  - processElectronicOrder:
      do:
        - validatePayment: {}
        - fulfillOrder: {}
        then: exit
  - processPhysicalOrder:
      branch:
        compete: true
        on:  # `multitask`, `parallelize` or `fork`
          - checkInventory: {}
          - packItems: {}
          - scheduleShipping: {}
            then: exit
  - handleUnknownOrderType: #...
```

I think you are mixing some proposals together because now here:

```yaml
 - processElectronicOrder:
      do:
        - validatePayment: {}
        - fulfillOrder: {}
        then: exit
```

`do` appears as a task which was only discusses in the sense of renaming `execute` to `do`

**Comment by matthias-pichler-warrify** [comment]
But it seems that on this one we are already very close to a consensus:

> 1. To make `do` a (sequential) array everywhere

that includes the main flow, `for` and `catch`. Is there still another way to then define sequential actions? Because it leaves us with no alternative for `try` and `before` & `after` in extensions. So either need to make these arrays as well or have some `execute/sequentially` alternative ... @JBBianchi proposed `do` itself:

 ```yaml
do:
  - processOrder:
      switch:
        - case1:
            when: .orderType == "electronic"
            then: processElectronicOrder
        - case2:
            when: .orderType == "physical"
            then: processPhysicalOrder
        - default:
            then: handleUnknownOrderType
  - processElectronicOrder:
      do: # do instead of execute:sequentially
        - validatePayment: {}
        - fulfillOrder: {}
      then: exit
  - processPhysicalOrder: #...
  - handleUnknownOrderType: #...
```

2. To get rid of `execute/sequentially`

that makes absolute sense to me because in my head "doing a set of tasks in order" should be the default mode for a workflow anyway. We just need to decide on how to substitute all its different use-cases, namely in `try`, `before`, `after` and nested tasks.

4. Create a new task to "parallelize" tasks

that also makes sense and we already have a couple of good ideas how to structure and name it (`fork`, `branch`, `multitask`, `parallelize`) 👍

**Comment by cdavernas** [comment]
> that includes the main flow, for and catch. Is there still another way to then define sequential actions? Because it leaves us with no alternative for try and before & after in extensions. So either need to make these arrays as well or have some execute/sequentially alternative

I think it should also apply to `try`, `before` and `after`. IMHO there's no reason a user should be not be able to use top level switch in those cases.

**Comment by cdavernas** [comment]
> agreed. But what should we do about nested sequential tasks? like @JBBianchi tried to show here:

Exactly what JB proposed I guess: basically use `do` to define a sequential composite task.

So:

1. Make do a (sequential) task array everywhere
2. Make `try`, `before` and `after` a (sequential) task array
3. Create a new `do` task, used to define a nested (sequential) task array
4. Create a new `fork`/`branch`/`multitask`/`parallelize` task, used to define a parallel (and possibly competing) task array

**Comment by fjtirado** [comment]
A  variant will be to use do: both for sequential and concurrency, as proposed here https://github.com/serverlessworkflow/specification/issues/894#issuecomment-2160183420.

**Comment by cdavernas** [comment]
> @cdavernas A variant for 4) will be to use do: both for sequential and concurrency, as proposed here https://github.com/serverlessworkflow/specification/issues/894#issuecomment-2160183420.

Nice idea! I'm however not a fan of the suggestion, for multiple reasons:

- It dois made out of two words
- It is not a verb, and does not respect imperative tense
- As @matthias-pichler-warrify mentionned on another issue, I would avoid having qualified properties at top level (ex: document/use/evaluate/do/onParallel <= feels off, and inconsistent)

**Comment by matthias-pichler-warrify** [comment]
> > @cdavernas A variant for 4) will be to use do: both for sequential and concurrency, as proposed here [#894 (comment)](https://github.com/serverlessworkflow/specification/issues/894#issuecomment-2160183420).
> 
> Nice idea! I'm however not a fan of the suggestion, for multiple reasons:
> 
> * It is made out of two words
> * It is not a verb, and does not respect imperative tense
> * As @matthias-pichler-warrify mentionned on another issue, I would avoid having qualified properties at top level (ex: document/use/evaluate/do/onParallel <= the later feels off, and inconsistent)

I am also more a fan of a dedicated task type for parallel execution since 

- it's what I (and therefore potentially others) are used to from other workflow languages
- concurrency (as any programmer knows) can be tough and making it _more_ explicit can't hurt
- it offers a nice place to add options (such as `compete` or `threads` etc) which do not make sense for sequential execution

**Comment by fjtirado** [comment]
I came from an OO world, so let me try to justfy why I feel adding a property to same object where do: is defined  is a good alternative to some new keyword that reflect the concept of do concurrently. 
I think we agree both do: and concurrently: (Im using the keyword for conveninece, we can use a different name) are array of task. 
do: is an array of task that is executed sequentially
concurrently: is an array of tas that is executed concurrently (concurrency might be configured)
Therefore it is not crazy to consider concurrently: an extension of do: and rather than redering it as a different keyword, render both as the same object with different internal state. 
In Json schema, that means defining a new object that contains do: as array of task and optional properties that setups how this array os task is run. Sequenatially if there is not any extra property. And concurrently if we write the addional property. 
That approach allow us, eventually, to add new properties that qualifies do: (the array of tasks) that are both shared by sequential and concurrent execution (we feel there are none, but we might be wrong)
The optional property onParallel: (we already have properties that are not vebs, like with, as...) is the place to add properties that are specific for concurrent execution
If we find out another way to execute the task (for example onReverse:) we just add the new optional property.

**Comment by fjtirado** [comment]
Anyway, I guess we are taking this approach

```
concurrently (or any other word):
   compete:
   #any additional property like compete
   branch (or some other word to write the array): 
      - task 1: 
     - task 2: 
```
That also works for me (my previous message was just for the record ;))

**Comment by cdavernas** [comment]
I'm myself an OO dev, I understand your point of view. 

My preference still remains the same. I do not like having the workflow becoming some kind of a do task with additional properties. I do not like having a non verb top level property. I also feel, as said by @matthias-pichler-warrify, that the proposed solution is much more intuitive. It is also much more consistent: feature get its own container class.

@JBBianchi @matthias-pichler-warrify @ricardozanini

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
I agree we can do with a dedicated keyword (so in essence we have a keyword for sequential list of task and another keyword for parallel list of task)
As a not related discussion, just for our sharing understanding, more than ordered and unordered, what distinguish a List (which is indeed fundamentally ordered) from a Set (note that SortedSet  is still a Set and sorting implies an order) in Java is that a List can be accessed by position (or in other words, any item in the list, at a certain moment, has a well known position identified by an index) while there is not such concept of  position in a Set (and also that any element in the Set is unique), even when the Set is ordered. 
To be honest, I prefer the STL approach, which distinguish between sequence containers (Java List) and associative containers (Java Set and Map).

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
I made some research regarding json schema in order to see if it was feasible to write documents like 
```
document:
  dsl: 1.0.0-alpha1
  namespace: default
  name: composite-sequential
  version: 1.0.0
do:
  set:
    colors: ${ .colors + ["red"] }
```
or

```

document:
  dsl: 1.0.0-alpha1
  namespace: default
  name: composite-sequential
  version: 1.0.0
do:
  sequentially:
  - setRed:
      set:
        colors: ${ .colors + ["red"] }
  - setGreen:
      set:
        colors: ${ .colors + ["green"] }
  - setBlue:
      set:
        colors: ${ .colors + ["blue"] }
```

So, removing "do" from top level and just leave there the "task" object as it is currently it is the schema, and it seems to work (Im bringing up here because I believe do, in the modified form we are discussing, should be a task. And the top level should refer to either any task or just the do task)

![image](https://github.com/serverlessworkflow/specification/assets/65240126/e716d498-3150-4ec0-bee6-bca17d51f3ff)

Basically, its removing do and adding task
```
oneOf:
- "$ref": "#/$defs/task"
required:
- document

```
I include the whole schema I used for reference
```

"$id": https://serverlessworkflow.io/schemas/1.0.0-alpha1/workflow.json
"$schema": http://json-schema.org/draft-07/schema
description: Serverless Workflow DSL - Workflow Schema
type: object
properties:
  document:
    type: object
    properties:
      dsl:
        type: string
        pattern: "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"
        description: The version of the DSL used by the workflow.
      namespace:
        type: string
        pattern: "^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$"
        description: The workflow's namespace.
      name:
        type: string
        pattern: "^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$"
        description: The workflow's name.
      version:
        type: string
        pattern: "^(0|[1-9]\\d*)\\.(0|[1-9]\\d*)\\.(0|[1-9]\\d*)(?:-((?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\\.(?:0|[1-9]\\d*|\\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\\+([0-9a-zA-Z-]+(?:\\.[0-9a-zA-Z-]+)*))?$"
        description: The workflow's semantic version.
      title:
        type: string
        description: The workflow's title.
      summary:
        type: string
        description: The workflow's Markdown summary.
      tags:
        type: object
        description: A key/value mapping of the workflow's tags, if any.
        additionalProperties: true
    required:
    - dsl
    - namespace
    - name
    - version
    description: Documents the workflow
  input:
    "$ref": "#/$defs/input"
    description: Configures the workflow's input.
  use:
    type: object
    properties:
      authentications:
        type: object
        additionalProperties:
          "$ref": "#/$defs/authenticationPolicy"
        description: The workflow's reusable authentication policies.
      errors:
        type: object
        additionalProperties:
          "$ref": "#/$defs/error"
        description: The workflow's reusable errors.
      extensions:
        type: array
        items:
          type: object
          minProperties: 1
          maxProperties: 1
          additionalProperties:
            "$ref": "#/$defs/extension"
        description: The workflow's extensions.
      functions:
        type: object
        additionalProperties:
          "$ref": "#/$defs/task"
        description: The workflow's reusable functions.
      retries:
        type: object
        additionalProperties:
          "$ref": "#/$defs/retryPolicy"
        description: The workflow's reusable retry policies.
      secrets:
        type: array
        items:
          type: string
        description: The workflow's secrets.
    description: Defines the workflow's reusable components.
  timeout:
    "$ref": "#/$defs/timeout"
    description: The workflow's timeout configuration, if any.
  output:
    "$ref": "#/$defs/output"
    description: Configures the workflow's output.
  schedule:
    type: object
    properties:
      every:
        "$ref": "#/$defs/duration"
        description: Specifies the duration of the interval at which the workflow
          should be executed.
      cron:
        type: string
        description: Specifies the schedule using a cron expression, e.g., '0 0 *
          * *' for daily at midnight."
      after:
        "$ref": "#/$defs/duration"
        description: Specifies a delay duration that the workflow must wait before
          starting again after it completes.
      'on':
        "$ref": "#/$defs/eventConsumptionStrategy"
        description: Specifies the events that trigger the workflow execution.
    description: Schedules the workflow
"$defs":
  task:
    type: object
    properties:
      input:
        "$ref": "#/$defs/input"
        description: Configure the task's input.
      output:
        "$ref": "#/$defs/output"
        description: Configure the task's output.
      timeout:
        "$ref": "#/$defs/timeout"
        description: The task's timeout configuration, if any.
      then:
        "$ref": "#/$defs/flowDirective"
        description: The flow directive to be performed upon completion of the task.
    oneOf:
    - "$ref": "#/$defs/callTask"
    - "$ref": "#/$defs/compositeTask"
    - "$ref": "#/$defs/emitTask"
    - "$ref": "#/$defs/forTask"
    - "$ref": "#/$defs/listenTask"
    - "$ref": "#/$defs/raiseTask"
    - "$ref": "#/$defs/runTask"
    - "$ref": "#/$defs/setTask"
    - "$ref": "#/$defs/switchTask"
    - "$ref": "#/$defs/tryTask"
    - "$ref": "#/$defs/waitTask"
  callTask:
    type: object
    oneOf:
    - properties:
        call:
          type: string
          const: asyncapi
        with:
          type: object
          properties:
            document:
              "$ref": "#/$defs/externalResource"
              description: The document that defines the AsyncAPI operation to call.
            operationRef:
              type: string
              description: A reference to the AsyncAPI operation to call.
            server:
              type: string
              description: A a reference to the server to call the specified AsyncAPI
                operation on. If not set, default to the first server matching the
                operation's channel.
            message:
              type: string
              description: The name of the message to use. If not set, defaults to
                the first message defined by the operation.
            binding:
              type: string
              description: The name of the binding to use. If not set, defaults to
                the first binding defined by the operation.
            payload:
              type: object
              description: The payload to call the AsyncAPI operation with, if any.
            authentication:
              description: The authentication policy, if any, to use when calling
                the AsyncAPI operation.
              oneOf:
              - "$ref": "#/$defs/authenticationPolicy"
              - type: string
          required:
          - document
          - operationRef
          description: Defines the AsyncAPI call to perform.
      required:
      - call
      - with
    - properties:
        call:
          type: string
          const: grpc
        with:
          type: object
          properties:
            proto:
              "$ref": "#/$defs/externalResource"
              description: The proto resource that describes the GRPC service to call.
            service:
              type: object
              properties:
                name:
                  type: string
                  description: The name of the GRPC service to call.
                host:
                  type: string
                  pattern: "^[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?$"
                  description: The hostname of the GRPC service to call.
                port:
                  type: integer
                  min: 0
                  max: 65535
                  description: The port number of the GRPC service to call.
                authentication:
                  description: The endpoint's authentication policy, if any.
                  oneOf:
                  - "$ref": "#/$defs/authenticationPolicy"
                  - type: string
              required:
              - name
              - host
            method:
              type: string
              description: The name of the method to call on the defined GRPC service.
            arguments:
              type: object
              additionalProperties: true
              description: The arguments, if any, to call the method with.
          required:
          - proto
          - service
          - method
          description: Defines the GRPC call to perform.
      required:
      - call
      - with
    - properties:
        call:
          type: string
          const: http
        with:
          type: object
          properties:
            method:
              type: string
              description: The HTTP method of the HTTP request to perform.
            endpoint:
              description: The HTTP endpoint to send the request to.
              oneOf:
              - "$ref": "#/$defs/endpoint"
              - type: string
                format: uri-template
            headers:
              type: object
              description: A name/value mapping of the headers, if any, of the HTTP
                request to perform.
            body:
              description: The body, if any, of the HTTP request to perform.
            output:
              type: string
              enum:
              - raw
              - content
              - response
              description: The http call output format. Defaults to 'content'.
          required:
          - method
          - endpoint
          description: Defines the HTTP call to perform.
      required:
      - call
      - with
    - properties:
        call:
          type: string
          const: openapi
        with:
          type: object
          properties:
            document:
              "$ref": "#/$defs/externalResource"
              description: The document that defines the OpenAPI operation to call.
            operationId:
              type: string
              description: The id of the OpenAPI operation to call.
            parameters:
              type: object
              additionalProperties: true
              description: A name/value mapping of the parameters of the OpenAPI operation
                to call.
            authentication:
              description: The authentication policy, if any, to use when calling
                the OpenAPI operation.
              oneOf:
              - "$ref": "#/$defs/authenticationPolicy"
              - type: string
            output:
              type: string
              enum:
              - raw
              - content
              - response
              description: The http call output format. Defaults to 'content'.
          required:
          - document
          - operationId
          description: Defines the OpenAPI call to perform.
      required:
      - call
      - with
    - properties:
        call:
          type: string
          not:
            enum:
            - asyncapi
            - grpc
            - http
            - openapi
          description: The name of the function to call.
        with:
          type: object
          additionalProperties: true
          description: A name/value mapping of the parameters, if any, to call the
            function with.
      required:
      - call
  compositeTask:
    type: object
    required:
    - do
    description: Serves as a pivotal orchestrator within workflow systems, enabling
      the seamless integration and execution of multiple subtasks to accomplish complex
      operations
    properties:
      execute:
        type: object
        description: Configures the task execution strategy to use
        oneOf:
        - required:
          - concurrently
          properties:
            concurrently:
              description: A list of the tasks to perform concurrently.
              type: array
              minItems: 2
              items:
                type: object
                minProperties: 1
                maxProperties: 1
                additionalProperties:
                  "$ref": "#/$defs/task"
            compete:
              description: Indicates whether or not the concurrent tasks are racing
                against each other, with a single possible winner, which sets the
                composite task's output.
              type: boolean
              default: false
        - required:
          - sequentially
          properties:
            sequentially:
              description: A list of the tasks to perform sequentially.
              type: array
              minItems: 2
              items:
                type: object
                minProperties: 1
                maxProperties: 1
                additionalProperties:
                  "$ref": "#/$defs/task"
  emitTask:
    type: object
    properties:
      emit:
        type: object
        properties:
          event:
            type: object
            properties:
              id:
                type: string
                description: The event's unique identifier
              source:
                type: string
                format: uri
                description: Identifies the context in which an event happened
              type:
                type: string
                description: This attribute contains a value describing the type of
                  event related to the originating occurrence.
              time:
                type: string
                format: date-time
              subject:
                type: string
              datacontenttype:
                type: string
                description: Content type of data value. This attribute enables data
                  to carry any type of content, whereby format and encoding might
                  differ from that of the chosen event format.
              dataschema:
                type: string
                format: uri
            required:
            - source
            - type
            additionalProperties: true
        required:
        - event
    required:
    - emit
    description: Allows workflows to publish events to event brokers or messaging
      systems, facilitating communication and coordination between different components
      and services.
  flowDirective:
    additionalProperties: false
    anyOf:
    - type: string
      enum:
      - continue
      - exit
      - end
      default: continue
    - type: string
  forTask:
    type: object
    properties:
      for:
        type: object
        properties:
          each:
            type: string
            description: The name of the variable used to store the current item being
              enumerated.
            default: item
          in:
            type: string
            description: A runtime expression used to get the collection to enumerate.
          at:
            type: string
            description: The name of the variable used to store the index of the current
              item being enumerated.
            default: index
        required:
        - in
      while:
        type: string
        description: A runtime expression that represents the condition, if any, that
          must be met for the iteration to continue.
      do:
        "$ref": "#/$defs/task"
    description: Allows workflows to iterate over a collection of items, executing
      a defined set of subtasks for each item in the collection. This task type is
      instrumental in handling scenarios such as batch processing, data transformation,
      and repetitive operations across datasets.
    required:
    - for
    - do
  listenTask:
    type: object
    properties:
      listen:
        type: object
        properties:
          to:
            "$ref": "#/$defs/eventConsumptionStrategy"
            description: Defines the event(s) to listen to.
        required:
        - to
    required:
    - listen
    description: Provides a mechanism for workflows to await and react to external
      events, enabling event-driven behavior within workflow systems.
  raiseTask:
    type: object
    properties:
      raise:
        type: object
        properties:
          error:
            "$ref": "#/$defs/error"
            description: Defines the error to raise.
        required:
        - error
    required:
    - raise
    description: Intentionally triggers and propagates errors.
  runTask:
    type: object
    properties:
      run:
        type: object
        oneOf:
        - properties:
            container:
              type: object
              properties:
                image:
                  type: string
                  description: The name of the container image to run.
                command:
                  type: string
                  description: The command, if any, to execute on the container
                ports:
                  type: object
                  description: The container's port mappings, if any.
                volumes:
                  type: object
                  description: The container's volume mappings, if any.
                environment:
                  type: object
                  description: A key/value mapping of the environment variables, if
                    any, to use when running the configured process.
              required:
              - image
          required:
          - container
          description: Enables the execution of external processes encapsulated within
            a containerized environment.
        - properties:
            script:
              type: object
              properties:
                language:
                  type: string
                  description: The language of the script to run.
                environment:
                  type: object
                  additionalProperties: true
                  description: A key/value mapping of the environment variables, if
                    any, to use when running the configured process.
              oneOf:
              - properties:
                  code:
                    type: string
                required:
                - code
                description: The script's code.
              - properties:
                  source:
                    "$ref": "#/$defs/externalResource"
                description: The script's resource.
                required:
                - code
              required:
              - language
          required:
          - script
          description: Enables the execution of custom scripts or code within a workflow,
            empowering workflows to perform specialized logic, data processing, or
            integration tasks by executing user-defined scripts written in various
            programming languages.
        - properties:
            shell:
              type: object
              properties:
                command:
                  type: string
                  description: The shell command to run.
                arguments:
                  type: object
                  additionalProperties: true
                  description: A list of the arguments of the shell command to run.
                environment:
                  type: object
                  additionalProperties: true
                  description: A key/value mapping of the environment variables, if
                    any, to use when running the configured process.
              required:
              - command
          required:
          - shell
          description: Enables the execution of shell commands within a workflow,
            enabling workflows to interact with the underlying operating system and
            perform system-level operations, such as file manipulation, environment
            configuration, or system administration tasks.
        - properties:
            workflow:
              type: object
              properties:
                namespace:
                  type: string
                  description: The namespace the workflow to run belongs to.
                name:
                  type: string
                  description: The name of the workflow to run.
                version:
                  type: string
                  default: latest
                  description: The version of the workflow to run. Defaults to latest
                input:
                  type: object
                  additionalProperties: true
                  description: The data, if any, to pass as input to the workflow
                    to execute. The value should be validated against the target workflow's
                    input schema, if specified.
              required:
              - namespace
              - name
              - version
          required:
          - workflow
          description: Enables the invocation and execution of nested workflows within
            a parent workflow, facilitating modularization, reusability, and abstraction
            of complex logic or business processes by encapsulating them into standalone
            workflow units.
    required:
    - run
    description: Provides the capability to execute external containers, shell commands,
      scripts, or workflows.
  setTask:
    type: object
    properties:
      set:
        type: object
        minProperties: 1
        additionalProperties: true
        description: The data to set
    required:
    - set
    description: A task used to set data
  switchTask:
    type: object
    properties:
      switch:
        type: array
        minItems: 1
        items:
          type: object
          minProperties: 1
          maxProperties: 1
          additionalProperties:
            type: object
            properties:
              name:
                type: string
                description: The case's name.
              when:
                type: string
                description: A runtime expression used to determine whether or not
                  the case matches.
              then:
                "$ref": "#/$defs/flowDirective"
                description: The flow directive to execute when the case matches.
    required:
    - switch
    description: Enables conditional branching within workflows, allowing them to
      dynamically select different paths based on specified conditions or criteria
  tryTask:
    type: object
    properties:
      try:
        description: The task to perform.
        "$ref": "#/$defs/task"
      catch:
        type: object
        properties:
          errors:
            type: object
          as:
            type: string
            description: The name of the runtime expression variable to save the error
              as. Defaults to 'error'.
          when:
            type: string
            description: A runtime expression used to determine whether or not to
              catch the filtered error
          exceptWhen:
            type: string
            description: A runtime expression used to determine whether or not to
              catch the filtered error
          retry:
            "$ref": "#/$defs/retryPolicy"
            description: The retry policy to use, if any, when catching errors.
          do:
            description: The definition of the task to run when catching an error.
            "$ref": "#/$defs/task"
    required:
    - try
    - catch
    description: Serves as a mechanism within workflows to handle errors gracefully,
      potentially retrying failed tasks before proceeding with alternate ones.
  waitTask:
    type: object
    properties:
      wait:
        "$ref": "#/$defs/duration"
        description: The amount of time to wait.
    required:
    - wait
    description: Allows workflows to pause or delay their execution for a specified
      period of time.
  authenticationPolicy:
    type: object
    oneOf:
    - properties:
        basic:
          type: object
          properties:
            username:
              type: string
              description: The username to use.
            password:
              type: string
              description: The password to use.
          required:
          - username
          - password
      required:
      - basic
      description: Use basic authentication.
    - properties:
        bearer:
          type: object
          properties:
            token:
              type: string
              description: The bearer token to use.
          required:
          - token
      required:
      - bearer
      description: Use bearer authentication.
    - properties:
        oauth2:
          type: object
          properties:
            authority:
              type: string
              format: uri
              description: The URI that references the OAuth2 authority to use.
            grant:
              type: string
              description: The grant type to use.
            client:
              type: object
              properties:
                id:
                  type: string
                  description: The client id to use.
                secret:
                  type: string
                  description: The client secret to use, if any.
              required:
              - id
            scopes:
              type: array
              items:
                type: string
              description: The scopes, if any, to request the token for.
            audiences:
              type: array
              items:
                type: string
              description: The audiences, if any, to request the token for.
            username:
              type: string
              description: The username to use. Used only if the grant type is Password.
            password:
              type: string
              description: The password to use. Used only if the grant type is Password.
            subject:
              "$ref": "#/$defs/oauth2Token"
              description: The security token that represents the identity of the
                party on behalf of whom the request is being made.
            actor:
              "$ref": "#/$defs/oauth2Token"
              description: The security token that represents the identity of the
                acting party.
          required:
          - authority
          - grant
          - client
      required:
      - oauth2
      description: Use OAUTH2 authentication.
    description: Defines an authentication policy.
  oauth2Token:
    type: object
    properties:
      token:
        type: string
        description: The security token to use to use.
      type:
        type: string
        description: The type of the security token to use to use.
    required:
    - token
    - type
  duration:
    type: object
    minProperties: 1
    properties:
      days:
        type: integer
        description: Number of days, if any.
      hours:
        type: integer
        description: Number of days, if any.
      minutes:
        type: integer
        description: Number of minutes, if any.
      seconds:
        type: integer
        description: Number of seconds, if any.
      milliseconds:
        type: integer
        description: Number of milliseconds, if any.
    description: The definition of a duration.
  error:
    type: object
    properties:
      type:
        type: string
        format: uri
        description: A URI reference that identifies the error type.
      status:
        type: integer
        description: The status code generated by the origin for this occurrence of
          the error.
      instance:
        type: string
        format: uri
        description: A JSON Pointer used to reference the component the error originates
          from.
      title:
        type: string
        description: A short, human-readable summary of the error.
      detail:
        type: string
        description: A human-readable explanation specific to this occurrence of the
          error.
    required:
    - type
    - status
    - instance
  endpoint:
    type: object
    properties:
      uri:
        type: string
        format: uri-template
        description: The endpoint's URI.
      authentication:
        description: The authentication policy to use.
        oneOf:
        - "$ref": "#/$defs/authenticationPolicy"
        - type: string
    required:
    - uri
  eventConsumptionStrategy:
    type: object
    oneOf:
    - properties:
        all:
          type: array
          items:
            "$ref": "#/$defs/eventFilter"
          description: A list containing all the events that must be consumed.
      required:
      - all
    - properties:
        any:
          type: array
          items:
            "$ref": "#/$defs/eventFilter"
          description: A list containing any of the events to consume.
      required:
      - any
    - properties:
        one:
          "$ref": "#/$defs/eventFilter"
          description: The single event to consume.
      required:
      - one
  eventFilter:
    type: object
    properties:
      with:
        type: object
        minProperties: 1
        properties:
          id:
            type: string
            description: The event's unique identifier
          source:
            type: string
            description: Identifies the context in which an event happened
          type:
            type: string
            description: This attribute contains a value describing the type of event
              related to the originating occurrence.
          time:
            type: string
          subject:
            type: string
          datacontenttype:
            type: string
            description: Content type of data value. This attribute enables data to
              carry any type of content, whereby format and encoding might differ
              from that of the chosen event format.
          dataschema:
            type: string
        additionalProperties: true
        description: An event filter is a mechanism used to selectively process or
          handle events based on predefined criteria, such as event type, source,
          or specific attributes.
      correlate:
        type: object
        additionalProperties:
          type: object
          properties:
            from:
              type: string
              description: A runtime expression used to extract the correlation value
                from the filtered event.
            expect:
              type: string
              description: A constant or a runtime expression, if any, used to determine
                whether or not the extracted correlation value matches expectations.
                If not set, the first extracted value will be used as the correlation's
                expectation.
          required:
          - from
        description: A correlation is a link between events and data, established
          by mapping event attributes to specific data attributes, allowing for coordinated
          processing or handling based on event characteristics.
    required:
    - with
    description: An event filter is a mechanism used to selectively process or handle
      events based on predefined criteria, such as event type, source, or specific
      attributes.
  extension:
    type: object
    properties:
      extend:
        type: string
        enum:
        - call
        - composite
        - emit
        - for
        - listen
        - raise
        - run
        - set
        - switch
        - try
        - wait
        - all
        description: The type of task to extend.
      when:
        type: string
        description: A runtime expression, if any, used to determine whether or not
          the extension should apply in the specified context.
      before:
        description: The task to execute before the extended task, if any.
        "$ref": "#/$defs/task"
      after:
        description: The task to execute after the extended task, if any.
        "$ref": "#/$defs/task"
    required:
    - extend
    description: The definition of a an extension.
  externalResource:
    type: object
    properties:
      uri:
        type: string
        format: uri
        description: The endpoint's URI.
      authentication:
        description: The authentication policy to use.
        oneOf:
        - "$ref": "#/$defs/authenticationPolicy"
        - type: string
      name:
        type: string
        description: The external resource's name, if any.
    required:
    - uri
  input:
    type: object
    properties:
      schema:
        "$ref": "#/$defs/schema"
        description: The schema used to describe and validate the input of the workflow
          or task.
      from:
        type: string
        description: A runtime expression, if any, used to mutate and/or filter the
          input of the workflow or task.
    description: Configures the input of a workflow or task.
  output:
    type: object
    properties:
      schema:
        "$ref": "#/$defs/schema"
        description: The schema used to describe and validate the output of the workflow
          or task.
      as:
        type: string
        description: A runtime expression, if any, used to mutate and/or filter the
          output of the workflow or task.
    description: Configures the output of a workflow or task.
  export:
    type: object
    properties:
      schema:
        "$ref": "#/$defs/schema"
        description: The schema used to describe and validate the workflow context.
      as:
        type: string
        description: A runtime expression, if any, used to export the output data
          to the context.
    description: Set the content of the context.
  retryPolicy:
    type: object
    properties:
      when:
        type: string
        description: A runtime expression, if any, used to determine whether or not
          to retry running the task, in a given context.
      exceptWhen:
        type: string
        description: A runtime expression used to determine whether or not to retry
          running the task, in a given context.
      delay:
        "$ref": "#/$defs/duration"
        description: The duration to wait between retry attempts.
      backoff:
        type: object
        oneOf:
        - properties:
            constant:
              type: object
              description: The definition of the constant backoff to use, if any.
          required:
          - constant
        - properties:
            exponential:
              type: object
              description: The definition of the exponential backoff to use, if any.
          required:
          - exponential
        - properties:
            linear:
              type: object
              description: The definition of the linear backoff to use, if any.
          required:
          - linear
        description: The retry duration backoff.
      limit:
        type: object
        properties:
          attempt:
            type: object
            properties:
              count:
                type: integer
                description: The maximum amount of retry attempts, if any.
              duration:
                "$ref": "#/$defs/duration"
                description: The maximum duration for each retry attempt.
          duration:
            "$ref": "#/$defs/duration"
            description: The duration limit, if any, for all retry attempts.
        description: The retry limit, if any
      jitter:
        type: object
        properties:
          from:
            "$ref": "#/$defs/duration"
            description: The minimum duration of the jitter range
          to:
            "$ref": "#/$defs/duration"
            description: The maximum duration of the jitter range
        required:
        - from
        - to
        description: The parameters, if any, that control the randomness or variability
          of the delay between retry attempts.
    description: Defines a retry policy.
  schema:
    type: object
    properties:
      format:
        type: string
        default: json
        description: The schema's format. Defaults to 'json'. The (optional) version
          of the format can be set using `{format}:{version}`.
    oneOf:
    - properties:
        document:
          description: The schema's inline definition.
      required:
      - document
    - properties:
        resource:
          "$ref": "#/$defs/externalResource"
          description: The schema's external resource.
      required:
      - resource
    description: Represents the definition of a schema.
  timeout:
    type: object
    properties:
      after:
        "$ref": "#/$defs/duration"
        description: The duration after which to timeout.
    required:
    - after
    description: The definition of a timeout.
oneOf:
- "$ref": "#/$defs/task"
required:
- document
```

**Comment by matthias-pichler-warrify** [comment]
@fjtirado 
I never claimed it wasn't possible. With your modification of moving `do` into the composite task you could also write it as

```yaml
type: object
required:
  - document
properties:
  document: {}
  use: {}
  ...
allOf:
  - $ref: "#/$defs/task"
``` 

which given the semantics of JSON schema would be the equivalent but more idiomatic way of doing it. But I thought it was thoroughly discussed that we did not want to make the top level workflow a task itself, because with your schema the following is valid:

```json
{
  "document": {
    "dsl": "1.0.0-alpha1",
    "namespace": "default",
    "name": "composite-sequential",
    "version": "1.0.0"
  },
  "call": "http",
  "with": {
    "method": "get",
    "endpoint": ""
  }
}
```

which defines a call task at the top level.

**Comment by fjtirado** [comment]
@matthias-pichler-warrify I understood that it was not possible, sorry
Since in order to reuse do/concurrently (in the form we are discussing now) I think the easiest way is to declared them as task (so you can reference them in try/loop/...) and from the top level just allow do/concurrently

**Comment by ricardozanini** [comment]
> > agreed. But what should we do about nested sequential tasks?
> 
> Exactly what @JBBianchi proposed I guess: basically use `do` to define a sequential composite task.
> 
> So, to summarize for sake of brievety, here's my understanding of what's proposed:
> 
> 1. Make `do` a (sequential) task array everywhere
> 2. Make `try`, `before` and `after` a (sequential) task array
> 3. Create a new `do` task, used to define a nested (sequential) task array
> 4. Create a new `fork`/`branch`/`multitask`/`parallelize` task, used to define a parallel (and possibly competing) task array

Nice summary, I think we already reached a consensus then. 

@JBBianchi are you opening the PR since you opened the issue? 

@matthias-pichler-warrify @fjtirado I understand that you guys are discussing the implementation in the schema, but the idea has been settled, correct?

**Comment by fjtirado** [comment]
@ricardozanini Yes, Im discussing how to implement the proposal with two keywords, do: and other word for concurrently in the schema

**Comment by matthias-pichler-warrify** [comment]
@fjtirado 
In your understanding should it be possible to have a top level `concurrently`

```yaml
document: {}
concurrently:
  - task1: {}
  - task2: {}
```

vs 

```yaml
document: {}
do:
  - task1:
      concurrently:
        - task2: {}
        - task3: {}
```

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
Yes, I guess is reasonable to accept concurrently at top level, but if we want to prevent that, we can just refer the do task at top level and thats it)
Regarding try/catch/loop, I will just reference the task object there (so user can write single task or multi task with do or concurrently)

**Comment by cdavernas** [comment]
> Yes, I guess is reasonable to accept concurrently at top level, but if we want to prevent that, we can just refer the do task at top level

I'd far prefer a 'concurrent' task.

> Regarding try/catch/loop, I will just reference the task object there (so user can write single task or multi task with do or concurrently)

Doing so would forbid the user to use a `switch` with those constructs.

**Comment by matthias-pichler-warrify** [comment]
@fjtirado 

for the implementation in the schema it is relatively straightforward (IMO)

```yaml
$schema: ""
type: object
required:
  - document
properties: 
  document: {}
  do:
    type: array
    items:
      type: object
      minProperties: 1
      maxProperties: 1
      additionalProperties:
        $ref: "#/$defs/task"
$defs:
  task:
    type: object
    properties: {}
    oneOf: [...]
  forTask:
    type: object
    required:
      - for
      - do
    properties:
      for: {}
      do:
        type: array
        items:
          type: object
          minProperties: 1
          maxProperties: 1
          additionalProperties:
            $ref: "#/$defs/task"
  tryTask:
    type: object
    required:
      - try
    properties:
      try:
        type: array
        items:
          type: object
          minProperties: 1
          maxProperties: 1
          additionalProperties:
            $ref: "#/$defs/task"
  # similar for catch
  branchTask:
    type: object
    required:
      - branch
    properties:
      branch:
        type: array
        items:
          type: object
          minProperties: 1
          maxProperties: 1
          additionalProperties:
            $ref: "#/$defs/task"
```

this schema would allow for the following definition

```yaml
document: {}
do:
  - branchTask:
      branch:
        - task1: {}
        - task2: {}
  - switchTask:
      switch:
        - case1:
            when: "..."
            then: "forTask"
        - case2:
            when: "..."
            then: "tryTask"
  - forTask:
      for: {}
      do:
        - task3: {}
        - task4: {}
  - tryTask:
      try:
        - task5: {}
        - task6: {}
      catch:
        do:
          - task7: {}
          - task8: {}
```

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify Yes, I guess is reasonable to accept concurrently at top level, but if we want to prevent that, we can just refer the do task at top level 

I would not add `concurrently` at the top. And `do` should not be a single task.

> Regarding try/catch/loop, I will just reference the task object there (so user can write single task or multi task with do or concurrently)

These should IMO not refer to a task object but to a task array. Just as the top level `do`

**Comment by cdavernas** [comment]
100% agree with @matthias-pichler-warrify

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
I like that schema ;)

Just for claryfing what I have in mind, I was thinking

```
document: {}
do:
  - branchTask:
      branch:
        - task1: {}
        - task2: {}
  - switchTask:
      switch:
        - case1:
            when: "..."
            then: "forTask"
        - case2:
            when: "..."
            then: "tryTask"
  - forTask:
      for: {}
      do:
        - task3: {}
        - task4: {}
  - tryTask:
      try:
        do:
           - task5: {}
           - task6: {}
      catch:
         call: http
```
I was trying to maintain Charles original idea of not naming single task in try/catch/loop, but Im perfectly fine with your proposal.

**Comment by JBBianchi** [comment]
I think you last suggestion is indeed what seems the clearest @matthias-pichler-warrify

**Comment by ricardozanini** [comment]
@matthias-pichler-warrify can you open a PR?

**Comment by matthias-pichler-warrify** [comment]
@fjtirado 
if your goal is maximum schema reuse one could do

```yaml
$schema: ""
type: object
required:
  - document
allOf: 
  - $ref: "#/$defs/doTask"
properties: 
  document: {}
$defs:
  task:
    type: object
    properties: {}
    oneOf: [...]
  forTask:
    type: object
    allOf: 
      - $ref: "#/$defs/doTask"
    required:
      - for
    properties:
      for: {}
  tryTask:
    type: object
    required:
      - try
    properties:
      try:
        $ref: "#/$defs/doTask/properties/do"
  # similar for catch
  branchTask:
    type: object
    required:
      - branch
    properties:
      branch:
        $ref: "#/$defs/doTask/properties/do"
  doTask:
    type: object
    required:
      - do
    properties:
      do:
        type: array
        items:
          type: object
          minProperties: 1
          maxProperties: 1
          additionalProperties:
            $ref: "#/$defs/task"
```

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify can you open a PR?

👍 sure thing

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
My goal was to allow not forcing naming of single task, but, again, I have not issue with naming all tasks (if you want single task, you just write an array of one).

**Comment by fjtirado** [comment]
@matthias-pichler-warrify I was thinking in 
```
$schema: ""
type: object
required:
  - document
properties: 
  document: {}
   oneOf:
      - $ref: "#/$defs/doTask"
      - $ref: "#/$defs/branchTask"
$defs:
  task:
    type: object
    properties: {}
    oneOf: [...]
  taskList:
    type: array
    items:
      type: object
      minProperties: 1
      maxProperties: 1
      additionalProperties:
        $ref: "#/$defs/task"
  forTask:
    type: object
    required:
      - for
      - do
    properties:
      for: {}
      do:
          - $ref: "#/$defs/task"
  tryTask:
    type: object
    required:
      - try
    properties:
      try:
        oneOf:
          - $ref: "#/$defs/task"
          - $ref: "#/$defs/taskList"
  # similar for catch
  branchTask:
    type: object
    required:
      - branch
    properties:
      branch:
        $ref: "#/$defs/taskList"
  doTask:
    type: object
    required:
      - do
    properties:
      do:
        $ref: "#/$defs/taskList"
```

And then you write
```

document: {}
do:
  - branchTask:
      branch:
        - task1: {}
        - task2: {}
  - switchTask:
      switch:
        - case1:
            when: "..."
            then: "forTask"
        - case2:
            when: "..."
            then: "tryTask"
  - forTask:
      for: {}
      do:
        - task3: {}
        - task4: {}
  - tryTask:
      try:
        do:
           - task5: {}
           - task6: {}
      catch:
         call: http
```

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify I was thinking in
> 
> ```
> $schema: ""
> type: object
> required:
>   - document
> properties: 
>   document: {}
>    oneOf:
>       - $ref: "#/$defs/doTask"
>       - $ref: "#/$defs/branchTask"
> $defs:
>   task:
>     type: object
>     properties: {}
>     oneOf: [...]
>   taskList:
>     type: array
>     items:
>       type: object
>       minProperties: 1
>       maxProperties: 1
>       additionalProperties:
>         $ref: "#/$defs/task"
>   forTask:
>     type: object
>     required:
>       - for
>       - do
>     properties:
>       for: {}
>       do:
>           - $ref: "#/$defs/task"
>   tryTask:
>     type: object
>     required:
>       - try
>     properties:
>       try:
>          $ref: "#/$defs/task"
>   # similar for catch
>   branchTask:
>     type: object
>     required:
>       - branch
>     properties:
>       branch:
>         $ref: "#/$defs/taskList"
>   doTask:
>     type: object
>     required:
>       - do
>     properties:
>       do:
>         $ref: "#/$defs/taskList"
> ```
> 
> And then you write
> 
> ```
> 
> document: {}
> do:
>   - branchTask:
>       branch:
>         - task1: {}
>         - task2: {}
>   - switchTask:
>       switch:
>         - case1:
>             when: "..."
>             then: "forTask"
>         - case2:
>             when: "..."
>             then: "tryTask"
>   - forTask:
>       for: {}
>       do:
>         - task3: {}
>         - task4: {}
>   - tryTask:
>       try:
>         do:
>            - task5: {}
>            - task6: {}
>       catch:
>          call: http
> ```

except it would be 

```yaml
for:
  do:
    do:
      - taks1: {}
      - taks2: {}
```

with your schema

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
I removed the do with a latter edition (we should discuss this on slack, which is more agile, not here)

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify I removed the do with a latter edition (we should discuss this on slack, which is more agile, not here)
> 
> with that you can write
> 
> ```
> - forTask:
>       for: {}
>       call: http
> ```

how do I now nest loops?

**Comment by matthias-pichler-warrify** [comment]
@fjtirado I believe we are going in circles ... all this was dicussed before

**Comment by fjtirado** [comment]
@matthias-pichler-warrify 
Since forTask is a task and forTask accept any task, you can nest loops.

**Comment by fjtirado** [comment]
@matthias-pichler-warrify I believe we need a quick chat on slack or zulip, How can I talk with you more fluently?

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify Since forTask is a task and forTask accept any task, you can nest loops.

show me a valid yaml instance of this:


it would need to be:

```yaml
- forTask:
   for: {}
   do:
     - forTask:
         for: {}
         call: http
```

**Comment by matthias-pichler-warrify** [comment]
> @matthias-pichler-warrify I believe we need a quick chat on slack or zulip, How can I talk with you more fluently?

Slack but how do I join the workspace?

**Comment by JBBianchi** [comment]
> > @matthias-pichler-warrify I believe we need a quick chat on slack or zulip, How can I talk with you more fluently?
> 
> Slack but how do I join the workspace?

It's the CNCF slack workspace, channel #serverless-workflow (https://communityinviter.com/apps/cloud-native/cncf I think ?)

**Comment by fjtirado** [comment]
Ok, after discussing it in slack. 
We rule out embedding task definition in for/try becuase of possible colision between task properties (although it was clarified that nested loops are allowed in that proposal ;)) and because of the swith problem (which is anyway a semantic one with arrays) 
As a consequence of that, there is no room for unnamed single task any longer. 
Therefore, lets implement https://github.com/serverlessworkflow/specification/issues/894#issuecomment-2161051207. 
We also discuss the possibility of having a slightly modified version of branch at top level, @matthias-pichler-warrify will give that a thought and, if needed, open a new issue after completing the PR associated with this one.

**Comment by matthias-pichler-warrify** [comment]
then I will get started with the PR 👍

**Comment by ricardozanini** [comment]
Thank you @matthias-pichler-warrify @fjtirado 🙏

**Comment by matthias-pichler-warrify** [comment]
we actually still need to pick a name from: 

> Create a new `fork`/`branch`/`multitask`/`parallelize` task, used to define a parallel (and possibly competing) task array

**Comment by JBBianchi** [comment]
> we actually still need to pick a name from:
> 
> > Create a new `fork`/`branch`/`multitask`/`parallelize` task, used to define a parallel (and possibly competing) task array

I have a little preference for `fork` or `branch` but I don't have a strong opinion about it.

**Comment by fjtirado** [comment]
@JBBianchi 
Same here

**Comment by matthias-pichler-warrify** [comment]
> > we actually still need to pick a name from:
> > > Create a new `fork`/`branch`/`multitask`/`parallelize` task, used to define a parallel (and possibly competing) task array
> 
> I have a little preference for `fork` or `branch` but I don't have a strong opinion about it.

I also prefer `fork` or `branch` 👍

**Comment by cdavernas** [comment]
As said by @matthias-pichler-warrify, both Amazon Step Functions and Google Cloud Workflows use the term "branches" to identify the collections of tasks to run in parallel, so I guess `branch` is the most adequate term.

On a side note for a plausible next good issue is that the branch task should define an array of `branch` key/value definitions, instead of an array of `task` key/value definitions. That would allow running branches in parallel, instead of actions, which I believe is more consistent with what is being proposed here, and would allow defining branch-specific properties.

**Comment by matthias-pichler-warrify** [comment]
> As said by @matthias-pichler-warrify, both Amazon Step Functions and Google Cloud Workflows use the term "branches" to identify the collections of tasks to run in parallel, so I guess `branch` is the most adequate term, but I don't have any strong personnal preference.
> 
> On a side note for a plausible next good issue is that the branch task should define an array of `branch` key/value definitions, instead of an array of `task` key/value definitions. That would allow running branches in parallel, instead of tasks, which I believe is more consistent with what is being proposed here, and would allow defining branch-specific configuration.
> 
> ```yaml
> branch:
>   - branch1:
>       - setFoo:
>           set:
>             foo: bar
>   - branch2:
>       - setBar:
>           set:
>             bar: baz
> ```

interesting 🤔 that could make a lot of sense and is consistent with `switch` cases.

on another note the current proposal (I think) includes `branch.on` ... should I keep it 

```yaml
document: {}
do:
  - branchTask:
       branch:
         # compete goes here
         on:
           - task1: {}
           - task2: {}
```

**Comment by cdavernas** [comment]
> on another note the current proposal (I think) includes branch.on ... should I keep it ?

I would then suggest using the term `fork` for the `task`, and `branches` instead of `on`, which is in my world heavily tight to the concept of events.

```yaml
do:
  - branchTask:
      fork:
        branches:
          - task1: {}
          - task2: {}
```

**Comment by fjtirado** [comment]
@cdavernas As you mention since order is not relevant for concurrent, we migth write 
```

do:
  - branchTask:
      fork:
        compete: true
        branches:
          task1: {}
          task2: {}
```

**Comment by matthias-pichler-warrify** [comment]
> I'm ok with @cdavernas proposal with `fork/branches`.
> 
> @fjtirado suggestion is also valid but I think it can be a bit confusing for the user as task lists are an array everywhere and then in `fork/branches`, suddenly (even if it's justified), it's an object.

I also like `fork/branches`.

@fjtirado the suggestion is valid but a little inconsistent as tasks only appear as arrays now

**Comment by fjtirado** [comment]
@JBBianchi @matthias-pichler-warrify 
I agree with you that map is a bit inconsistent with array. No preference from my side, to be honest

**Comment by matthias-pichler-warrify** [comment]
#895 should be ready for review 👍
